packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -5647,6 +5652,137 @@ namespace Tpetra {
+    Kokkos::deep_copy(row_ptrs_beg, this->k_rowPtrs_);
+
+    const size_t N = (row_ptrs_beg.extent(0) == 0 ? 0 : row_ptrs_beg.extent(0) - 1);
+    row_ptrs_type row_ptrs_end("row_ptrs_end", N);
+
+    bool refill_num_row_entries = false;
+    if (this->k_numRowEntries_.extent(0) > 0) {
+      // Case 1: Unpacked storage
+      refill_num_row_entries = true;
+      auto num_row_entries = this->k_numRowEntries_;
+      Kokkos::parallel_for("Fill end row pointers", range_policy(0, N),
+        KOKKOS_LAMBDA(const size_t i){
+          row_ptrs_end(i) = row_ptrs_beg(i) + num_row_entries(i);
+        }
+      );
+
+    } else {
+      // mfh If packed storage, don't need row_ptrs_end to be separate allocation;
+      // could just have it alias row_ptrs_beg+1.
+      // Case 2: Packed storage
+      Kokkos::parallel_for("Fill end row pointers", range_policy(0, N),
+        KOKKOS_LAMBDA(const size_t i){
+          row_ptrs_end(i) = row_ptrs_beg(i+1);
+        }
+      );
+    }
+
+    using padding_type = Kokkos::UnorderedMap<LocalOrdinal, size_t, device_type>;
+    padCrsArrays<row_ptrs_type,indices_type,padding_type>(
+        row_ptrs_beg, row_ptrs_end, indices, padding);
+
+    if (refill_num_row_entries) {
+      auto num_row_entries = this->k_numRowEntries_;
+      Kokkos::parallel_for("Fill num entries", range_policy(0, N),
+        KOKKOS_LAMBDA(const size_t i){
+          num_row_entries(i) = row_ptrs_end(i) - row_ptrs_beg(i);
+        }
+      );
+    }
+    this->k_rowPtrs_ = row_ptrs_beg;
+    this->k_gblInds1D_ = indices;
+  }
+
+  template <class LocalOrdinal, class GlobalOrdinal, class Node>
+  Kokkos::UnorderedMap<LocalOrdinal, size_t, typename Node::device_type>
+  CrsGraph<LocalOrdinal, GlobalOrdinal, Node>::
+  computeCrsPadding (const RowGraph<LocalOrdinal,GlobalOrdinal,Node>& source,
+                     size_t numSameIDs,
+                     const Teuchos::ArrayView<const LocalOrdinal> &permuteToLIDs,
+                     const Teuchos::ArrayView<const LocalOrdinal> &permuteFromLIDs)
+  {
+    using LO = LocalOrdinal;
+    using GO = GlobalOrdinal;
+    using execution_space = typename device_type::execution_space;
+    const char tfecfFuncName[] = "computeCrsPadding";
+
+    // Resize row pointers and indices to accommodate incoming data
+    execution_space::fence ();  // Make sure device sees changes made by host
+    const map_type& src_row_map = *(source.getRowMap());
+    using padding_type = Kokkos::UnorderedMap<LocalOrdinal, size_t, device_type>;
+    padding_type padding(numSameIDs+permuteFromLIDs.size());
+    for (LO tgtid=0; tgtid<static_cast<LO>(numSameIDs); ++tgtid) {
+      const GO srcgid = src_row_map.getGlobalElement(tgtid);
+      auto how_much_padding = source.getNumEntriesInGlobalRow(srcgid);
+      auto result = padding.insert(tgtid, how_much_padding);
+      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(result.failed(), std::runtime_error,
```

Comment:
FYI for later: When we thread parallelize this, we should later just `parallel_reduce` over an error code.  The only reason we need this not to be parallel for now, is because the source is a RowGraph and not necessarily a fill-complete CrsGraph.